### PR TITLE
[FRONTEND][TENSORFLOW] Enhance with left over patches from NNVM.

### PR DESCRIPTION
**Pull request by srkreddy1238**
commit 76188a43d062ecf87721f6672dad0d9f714efc6d
Author: Siva <sivar.b@huawei.com>
    [NNVM][TENSORFLOW] bugfix. (#2444)

commit 6737739c3753f3ebce96b3426de7cd0e546582fa
Author: Ashutosh Parkhi <ashutosh.parkhi@imgtec.com>
    [Tensorflow] Support for Crop (#2285)

commit f6c3f9976096de188ed0060b35a4266885513b98
Author: Alexey Romanov <alexey.v.romanov@gmail.com>
    [FRONTEND][TENSORFLOW] Use input shapes directly instead of 1-element lists (#2242)

commit e5d92e1b96c1fbc175be741f522c030f2d91a613
Author: Dominic Symes <36929632+dominicsymes@users.noreply.github.com>
    [FRONTEND][TENSORFLOW] Bugfix (#2326)

commit 00d509d480ae55def1abccc61742a163dadfab39
Author: Alexey Romanov <alexey.v.romanov@gmail.com>
    [FRONTEND][TENSORFLOW] Support Unstack and Split (#2105)

commit df9d3ad299d550e0f42504fedc149866a87e1c0b
Author: Siva <sivar.b@huawei.com>
    [FRONTEND][TENSORFLOW] Bugfix (#2267)

commit d1a0c901067e382f8e4d75a45ab2faeffb86622f
Author: Zhebin Jin <zhebin.jzb@alibaba-inc.com>
    [FRONTEND][TENSORFLOW]Add Split and realdiv op support (#2123)
    * Add Split and realdiv op support
    * Fix the pad calculation in the case of dilated convolution

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

### CODE REVIEW

**Comment by tqchen** [comment]
@kazum @Huyuwei please review this PR

**Comment by srkreddy1238** [comment]
Also @yongwww & @ashutoshparkhi welcome to review

File: nnvm/tests/python/frontend/tensorflow/test_forward.py [hunk 1]
```diff
@@ -1235,7 +1233,7 @@ def test_forward_rel_ops():
     test_forward_inception_v3()
     test_forward_inception_v1()
     test_forward_mobilenet()
-    test_forward_resnetv2()
+    #_test_forward_resnetv2()
```
Comment (kazum) [inline nnvm/tests/python/frontend/tensorflow/test_forward.py]: Why do you disable the resnetv2 test?

File: python/tvm/relay/frontend/tensorflow.py [hunk 2]
```diff
@@ -1348,15 +1480,29 @@ def from_tensorflow(self, graph, layout="NHWC", shape=None, outputs=None):
...
+                          or -1 in self._output_shapes[node.name][0]):
+                if isinstance(node_output, _expr.TupleWrapper):
+                    out_shapes = []
+                    for tuple_item in node_output:
+                        out_type = ir_pass.infer_type(tuple_item)
+                        out_shapes.append(get_const_tuple(out_type.checked_type.shape))
+                    self._output_shapes[node.name] = out_shapes
+                else:
+                    out_type = ir_pass.infer_type(node_output[0])
+                    self._output_shapes[node.name] = [get_const_tuple(out_type.checked_type.shape)]
```
Comment (kazum) [inline python/tvm/relay/frontend/tensorflow.py]: I guess the length of node_output is always 1, no?  If yes, we can create out_shapes in the same way as TupleWrapper.

File: nnvm/tests/python/frontend/tensorflow/test_forward.py (hunk 1 above)
Comment (srkreddy1238) [reply nnvm/tests/python/frontend/tensorflow/test_forward.py]: Resnet V2 was enabled with GPU target earlier.
Tensorflow on CI doesn't support GPU now. Also I see some strange error with GPU in local setup.

Hence disabled now, I will consider it later.

File: python/tvm/relay/frontend/tensorflow.py (hunk 2 above)
Comment (srkreddy1238) [reply python/tvm/relay/frontend/tensorflow.py]: good catch. thanks, I will change

File: nnvm/tests/python/frontend/tensorflow/test_forward.py (hunk 1 above)
Comment (srkreddy1238) [reply nnvm/tests/python/frontend/tensorflow/test_forward.py]: fixed this too.

File: nnvm/tests/python/frontend/tensorflow/test_forward.py [hunk 3]
```diff
@@ -240,9 +240,7 @@ def _test_convolution(tensor_in_sizes, filter_in_sizes,
 def test_forward_convolution():
     if is_gpu_available():
         _test_convolution([4, 176, 8, 8], [1, 1, 176, 32], [1, 1], [1, 1], 'SAME', 'NCHW')
-        _test_convolution([4, 19, 17, 17], [3, 3, 19, 19], [1, 1], [2, 2], 'VALID', 'NCHW')
         _test_convolution([4, 124, 17, 17], [1, 1, 124, 19], [1, 1], [1, 1], 'SAME', 'NCHW')
-        _test_convolution([4, 12, 17, 17], [3, 3, 12, 32], [1, 1], [2, 2], 'VALID', 'NCHW')
```
Comment (kazum) [inline nnvm/tests/python/frontend/tensorflow/test_forward.py]: Any reason why padding=='VALID' tests are removed?

File: python/tvm/relay/frontend/tensorflow.py [hunk 4]
```diff
@@ -752,6 +778,9 @@ def _transform_mask(stride_dim, ellipsis_mask):
                 pass
             else:
                 final_output.append(out_shape[gather_index])
+        # Prevent 0-dim tensors which are not accepted by nnvm
```
Comment (kazum) [inline python/tvm/relay/frontend/tensorflow.py]:
```suggestion
        # Prevent 0-dim tensors which are not accepted by Relay
```

File: python/tvm/relay/frontend/tensorflow.py [hunk 5]
```diff
@@ -1315,10 +1424,33 @@ def from_tensorflow(self, graph, layout="NHWC", shape=None, outputs=None):
...
+                        if isinstance(in_sym, _expr.TupleWrapper):
+                            tensor_slot = int(tensor_name[1]) if len(tensor_name) > 1 else 0
+                            in_sym = [in_sym[tensor_slot]]
+                            input_shape = self._output_shapes[node_name][tensor_slot]
+                        else:
+                            tensor_slot = 0
+                            input_shape = self._output_shapes[node_name][0]
+                        inputs.append(in_sym[0])
+                        input_shapes[in_sym[0]] = input_shape
+                        # This means the node is 1d in NNVM and 0d in TF.
```
Comment (kazum) [inline python/tvm/relay/frontend/tensorflow.py]:
```suggestion
                        # This means the node is 1d in Relay and 0d in TF.
```

File: nnvm/tests/python/frontend/tensorflow/test_forward.py (hunk 3 above)
Comment (srkreddy1238) [reply nnvm/tests/python/frontend/tensorflow/test_forward.py]: :+1:  enabled the test cases with fixes. 

After correcting GPU flag check function some cases become invalid.

Comment (srkreddy1238) [reply nnvm/tests/python/frontend/tensorflow/test_forward.py]: The GPU check flag fix in this PR caused some test cases invalid for tensorflow GPU.
Fixed all of them now :+1:

File: python/tvm/relay/frontend/tensorflow.py [hunk 6]
```diff
@@ -1267,15 +1369,22 @@ def from_tensorflow(self, graph, layout="NHWC", shape=None, outputs=None):
             elif shape and node.name in shape:
                 # Give priority to user argument.
                 self._output_shapes[node.name] = [shape[node.name]]
+            elif node.op == 'Placeholder':
```
[the remark this answers is not in the document]
Comment (srkreddy1238) [reply python/tvm/relay/frontend/tensorflow.py]: You are right. Removed now.

**Comment by tqchen** [comment]
@srkreddy1238 please send another push to re-trigger CI. @yongwww please https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

File: python/tvm/relay/frontend/tensorflow.py [hunk 7]
```diff
@@ -1315,10 +1424,33 @@ def from_tensorflow(self, graph, layout="NHWC", shape=None, outputs=None):
...
+                            input_shape = self._output_shapes[node_name][tensor_slot]
+                        else:
+                            tensor_slot = 0
+                            input_shape = self._output_shapes[node_name][0]
+                        inputs.append(in_sym[0])
+                        input_shapes[in_sym[0]] = input_shape
+                        # This means the node is 1d in Relay and 0d in TF.
+                        # See `_expand_dims_0d_aware`.
+                        if self._outputs_are_0d[node_name][tensor_slot] and input_shape:
+                            input_0d_mismatch.add(in_sym)
```
Comment (ashutoshparkhi) [inline python/tvm/relay/frontend/tensorflow.py]: Shouldn't this be in_sym[tensor_slot]?

Comment (srkreddy1238) [reply python/tvm/relay/frontend/tensorflow.py]: I think tensor_slot us already handled above.

Comment (ashutoshparkhi) [reply python/tvm/relay/frontend/tensorflow.py]: In my variant of inception_v3, in_sym comes as a list of Var instead of _expr.TupleWrapper.

Comment (yongwww) [reply python/tvm/relay/frontend/tensorflow.py]: @ashutoshparkhi will fix it in pr https://github.com/dmlc/tvm/pull/2830

**Pull request merged**